### PR TITLE
Add luacheck syntax checker

### PR DESCRIPTION
**Pull request by ptrv**
Adding `luackeck` to the test VM is very simple:

After lua52 is installed you need following steps
1. Download and unpack latest luacheck from https://github.com/mpeterv/luacheck/archive/0.10.0.tar.gz
2. Run `sudo lua install.lua /usr/local` inside luacheck source folder

Now the `luacheck` executable is in `/usr/local/bin` which is probably in `PATH`

### CODE REVIEW

**Comment by ptrv** [comment]
Added also the ansible part. Extended the lua language role to install luacheck.

File: flycheck.el [hunk 1]
```diff
@@ -6330,6 +6331,118 @@ See URL `http://lesscss.org'."
...
+                        ,(nth 5 flycheck-luacheck-std-globals-list))
+                 (const :tag "Globals of LuaJIT 2.0"
+                        ,(nth 6 flycheck-luacheck-std-globals-list))
+                 (const :tag "Intersection of globals of Lua 5.1, 5.2 and LuaJIT 2.0"
+                        ,(nth 7 flycheck-luacheck-std-globals-list))
+                 (const :tag "Union of globals of Lua 5.1, 5.2 and LuaJIT 2.0"
+                        ,(nth 8 flycheck-luacheck-std-globals-list))
+                 (const :tag "No standard globals"
+                        ,(nth 9 flycheck-luacheck-std-globals-list)))
+  :safe (lambda (value) (member value flycheck-luacheck-std-globals-list))
```
Comment (lunaryorn) [inline flycheck.el]: This is way too detailed and too complex. Please make this option a simple string.

File: flycheck.el [hunk 2]
```diff
@@ -6330,6 +6331,118 @@ See URL `http://lesscss.org'."
...
+            "--no-config"
+            (option "--std" flycheck-luacheck-std-globals)
+            (option-flag "--no-global" flycheck-luacheck-no-globals)
+            (option-flag "--no-unused" flycheck-luacheck-no-unused)
+            (option-flag "--no-redefined" flycheck-luacheck-no-redefined)
+            (option-flag "--no-unused-args" flycheck-luacheck-no-unused-args)
+            (option-flag "--no-unused-secondaries" flycheck-luacheck-no-unused-secondaries)
+            (option-flag "--no-unused-globals" flycheck-luacheck-no-unused-globals)
+            "--formatter" "plain"
+            "--codes"
```
Comment (lunaryorn) [inline flycheck.el]: What does `--codes` do?

File: flycheck.el [hunk 3]
```diff
@@ -6330,6 +6331,118 @@ See URL `http://lesscss.org'."
...
+                         (string :tag "Global name")))
+  :safe #'flycheck-string-list-p
+  :package-version '(flycheck . "0.23"))
+
+(flycheck-define-checker luacheck
+  "A Lua syntax checker using luacheck.
+
+See URL `https://github.com/mpeterv/luacheck'."
+  :command ("luacheck"
+            "--no-config"
```
Comment (lunaryorn) [inline flycheck.el]: What does this option do?

File: test/flycheck-test.el [hunk 4]
```diff
@@ -4423,11 +4423,42 @@ Why not:
...
    '(1 1 error "Unrecognised input" :checker less)))
 
-(flycheck-ert-def-checker-test lua lua nil
+(flycheck-ert-def-checker-test luacheck lua syntax-error
   (flycheck-ert-should-syntax-check
    "checkers/lua-syntax-error.lua" 'lua-mode
-   '(5 nil error "unfinished string near '\"oh no'"
-       :checker lua)))
+   '(5 7 error "unfinished string"
+       :checker luacheck)))
```
Comment (lunaryorn) [inline test/flycheck-test.el]: Why did you remove the test for the `lua` checker?

File: playbooks/roles/language-lua/tasks/luacheck.yml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+- name: Download luacheck
+  get_url: url=https://github.com/mpeterv/luacheck/archive/{{luacheck_version}}.zip
+           dest=/usr/src/luacheck-{{luacheck_version}}.zip
+- name: Extract luacheck
+  unarchive: src=/usr/src/luacheck-{{luacheck_version}}.zip copy=no
+             dest=/opt
+- name: Install luacheck
+  command: /usr/bin/lua /opt/luacheck-{{luacheck_version}}/install.lua /usr/local
+  args:
+    chdir: /opt/luacheck-{{luacheck_version}}
```
Comment (lunaryorn) [inline playbooks/roles/language-lua/tasks/luacheck.yml]: Just being curious: How long does it take to build?

File: flycheck.el (hunk 1 above)
Comment (ptrv) [inline flycheck.el]: Ok, I will simplify this

File: flycheck.el (hunk 2 above)
Comment (ptrv) [inline flycheck.el]: It adds the error code, e.g. `W212`. For errors there is no code, so this makes it possible to use the  `id` in the error format and distinct between errors and warnnings.

File: flycheck.el (hunk 3 above)
Comment (ptrv) [inline flycheck.el]: This option disables the option to load an external config file. I did not introduce the option of specifying an external config file on purpose because you could set the `luacheck` output format or remove support for `codes` which would make the syntax checker fail.

File: test/flycheck-test.el (hunk 4 above)
Comment (ptrv) [inline test/flycheck-test.el]: I haven't removed the test for the `lua` checker. It's just below the new `luacheck` tests

File: playbooks/roles/language-lua/tasks/luacheck.yml (hunk 5 above)
Comment (ptrv) [inline playbooks/roles/language-lua/tasks/luacheck.yml]: It's pretty fast because the `install.lua` script basically just copies the files to the right places without any building.

**Comment by ptrv** [comment]
Updated the commit according to your requests

File: flycheck.el (hunk 3 above)
Comment (lunaryorn) [inline flycheck.el]: @ptrv Add a config file nonetheless.  Other config files (e.g. Pylint) allow for the same, but we still provide them.

File: flycheck.el (hunk 2 above)
Comment (lunaryorn) [inline flycheck.el]: @ptrv Would you mind to add a small comment that explains the use of `--codes` here?  This would make it easier for me to understand what's going on later.

File: playbooks/roles/language-lua/tasks/luacheck.yml (hunk 5 above)
Comment (lunaryorn) [inline playbooks/roles/language-lua/tasks/luacheck.yml]: Cool, thanks

File: test/flycheck-test.el (hunk 4 above)
Comment (lunaryorn) [inline test/flycheck-test.el]: @ptrv Ah, sorry, I overlooked it.  Sorry for the noise.

File: flycheck.el (hunk 3 above)
Comment (ptrv) [inline flycheck.el]: @lunaryorn well, ok. I'll add the config file options. Should I remove some of the options then? I am not sure how the `luacheck` behaves when setting both. Guess I have to check the case having stuff both in the config and via arument

Comment (mpeterv) [inline flycheck.el]: @ptrv command-line options always override config.

Comment (ptrv) [inline flycheck.el]: @mpeterv Thanks for clarifying. The, specifying `--formatter` and `--codes` as arguments will never be overwritten from config.

Comment (lunaryorn) [inline flycheck.el]: @ptrv I'm not sure.  Generally, I prefer configuration files, since they can be shared among all contributors to a project and are independent of a specific editor or extension.  I think we only need to keep those options that the user is really likely to set individually.

**Comment by ptrv** [comment]
Simplified the checker by removing all options and adding the `config-file` and since command-line options override the config file options the checker output format cannot be overridden.

**Comment by ptrv** [comment]
@lunaryorn  Added requests and rebased onto master.

**Comment by lunaryorn** [comment]
@ptrv Thanks, merged!

**Comment by ptrv** [comment]
@lunaryorn Thanks for merging!
